File: vincoreai/13-templates/config_loader.js
/**
 * Vincore AI — Config Loader (Node.js)
 * ─────────────────────────────────────
 * Import this at the top of every build script.
 * It loads client_config.json, resolves document IDs from
 * the project code, and validates that no placeholder values
 * remain unfilled before a build runs.
 *
 * Usage in any build script:
 *   const C = require('./config_loader');
 *   run(C.client.business_name, ...)   // → actual client name
 */

const fs   = require('fs');
const path = require('path');

// ── Load raw config ──────────────────────────────────────────────
const configPath = path.join(__dirname, 'client_config.json');
if (!fs.existsSync(configPath)) {
  throw new Error(`client_config.json not found at ${configPath}. Create it before running any build script.`);
}

const raw = JSON.parse(fs.readFileSync(configPath, 'utf8'));

// ── Resolve document IDs from project code ────────────────────────
const code = raw.project.code;
const docs = {};
for (const [key, val] of Object.entries(raw.documents)) {
  docs[key] = val.replace('[PROJECT_CODE]', code);
}

// ── Convenience shortcuts ─────────────────────────────────────────
const C = {
  // Core sections
  vincore:    raw.vincore,
  client:     raw.client,
  project:    raw.project,
  commercial: raw.commercial,
  system:     raw.system,
  output:     raw.output,
  docs,

  // Frequently used aliases
  CLIENT:     raw.client.business_name,
  CONTACT:    raw.client.contact_name,
  EMAIL:      raw.client.email,
  PHONE:      raw.client.phone,
  LOCATION:   raw.client.location,
  WEBSITE:    raw.client.website,

  CODE:       raw.project.code,
  NAME:       raw.project.name,
  FEE:        raw.commercial.fee_total,
  DEPOSIT:    raw.commercial.fee_deposit,
  BALANCE:    raw.commercial.fee_balance,
  CURRENCY:   raw.commercial.currency_symbol,
  START:      raw.project.start_date,
  END:        raw.project.end_date,
  GOLIVE:     raw.project.go_live_date,
  MONTH:      raw.output.report_month,
  VERSION:    raw.output.version,
  OUT:        raw.output.dir,

  // Helper: load client logo (returns buffer)
  logo: () => {
    const lp = path.join(__dirname, raw.client.logo_path);
    if (!fs.existsSync(lp)) {
      console.warn(`⚠  Logo not found at ${lp}. Using placeholder.`);
      return null;
    }
    return fs.readFileSync(lp);
  },

  // Helper: output filename builder
  // e.g. C.filename('CDR', 'docx') → 'VAI-CLI-XXX-001-CDR_ClientDiscoveryReport_v1.0.docx'
  filename: (docKey, ext, label='') => {
    const id = docs[docKey] || docKey;
    const slug = label ? `_${label.replace(/\s+/g,'').replace(/[^a-zA-Z0-9]/g,'')}` : '';
    return `${id}${slug}_v${raw.output.version}.${ext}`;
  },
};

// ── Validate: warn about unfilled placeholders ─────────────────────
function findUnfilled(obj, path = '') {
  const issues = [];
  for (const [k, v] of Object.entries(obj)) {
    if (k.startsWith('_')) continue;
    const fullPath = path ? `${path}.${k}` : k;
    if (typeof v === 'string' && v.includes('[')) {
      issues.push(`  ⚠  ${fullPath}: "${v}"`);
    } else if (typeof v === 'object' && v !== null && !Array.isArray(v)) {
      issues.push(...findUnfilled(v, fullPath));
    }
  }
  return issues;
}

const issues = findUnfilled(raw);
if (issues.length > 0) {
  console.warn('\n┌─ client_config.json has unfilled placeholders ─────────────');
  issues.forEach(i => console.warn(i));
  console.warn('└────────────────────────────────────────────────────────────\n');
}

module.exports = C;

File: vincoreai/13-templates/config_loader.py
"""
Vincore AI — Config Loader (Python)
─────────────────────────────────────
Import this at the top of every Python build script.
Loads client_config.json and resolves document IDs.

Usage:
    from config_loader import C
    print(C['client']['business_name'])
    print(C['docs']['pmp'])
    print(C['CLIENT'])             # shortcut
"""

import json
import os
import sys
import warnings

# ── Load raw config ──────────────────────────────────────────────
_DIR = os.path.dirname(os.path.abspath(__file__))
_CONFIG_PATH = os.path.join(_DIR, 'client_config.json')

if not os.path.exists(_CONFIG_PATH):
    raise FileNotFoundError(
        f'client_config.json not found at {_CONFIG_PATH}. '
        'Create it before running any build script.'
    )

with open(_CONFIG_PATH, encoding='utf-8') as f:
    _raw = json.load(f)

# ── Resolve document IDs from project code ─────────────────────────
_code = _raw['project']['code']
docs = {
    key: val.replace('[PROJECT_CODE]', _code)
    for key, val in _raw['documents'].items()
}

# ── Build convenience object ───────────────────────────────────────
C = {
    # Full sections
    'vincore':    _raw['vincore'],
    'client':     _raw['client'],
    'project':    _raw['project'],
    'commercial': _raw['commercial'],
    'system':     _raw['system'],
    'output':     _raw['output'],
    'docs':       docs,

    # Flat shortcuts
    'CLIENT':   _raw['client']['business_name'],
    'CONTACT':  _raw['client']['contact_name'],
    'EMAIL':    _raw['client']['email'],
    'PHONE':    _raw['client']['phone'],
    'LOCATION': _raw['client']['location'],
    'WEBSITE':  _raw['client']['website'],

    'CODE':     _raw['project']['code'],
    'NAME':     _raw['project']['name'],
    'FEE':      _raw['commercial']['fee_total'],
    'DEPOSIT':  _raw['commercial']['fee_deposit'],
    'BALANCE':  _raw['commercial']['fee_balance'],
    'CURRENCY': _raw['commercial']['currency_symbol'],
    'START':    _raw['project']['start_date'],
    'END':      _raw['project']['end_date'],
    'GOLIVE':   _raw['project']['go_live_date'],
    'MONTH':    _raw['output']['report_month'],
    'VERSION':  _raw['output']['version'],
    'OUT':      _raw['output']['dir'],
}

# Helper: output filename builder
def make_filename(doc_key: str, ext: str, label: str = '') -> str:
    """e.g. make_filename('pmp','xlsx') → 'VAI-CLI-XXX-001-PMP_v1.0.xlsx'"""
    doc_id = docs.get(doc_key, doc_key)
    slug = ('_' + ''.join(c for c in label if c.isalnum())) if label else ''
    return f"{doc_id}{slug}_v{_raw['output']['version']}.{ext}"

C['filename'] = make_filename

# ── Validate: warn about unfilled placeholders ─────────────────────
def _check(obj, path=''):
    issues = []
    if isinstance(obj, dict):
        for k, v in obj.items():
            if k.startswith('_'):
                continue
            fp = f'{path}.{k}' if path else k
            issues.extend(_check(v, fp))
    elif isinstance(obj, str) and '[' in obj:
        issues.append(f'  ⚠  {path}: "{obj}"')
    return issues

_issues = _check(_raw)
if _issues:
    print('\n┌─ client_config.json has unfilled placeholders ─────────────')
    for i in _issues:
        print(i)
    print('└────────────────────────────────────────────────────────────\n')
